src/librustc_resolve/lib.rs

```diff
@@ -2697,20 +2699,50 @@ impl<'a> Resolver<'a> {
                         }
                         return (err, candidates);
                     },
-                    _ if ns == ValueNS && is_struct_like(def) => {
-                        if let Def::Struct(def_id) = def {
-                            if let Some((ctor_def, ctor_vis))
-                                    = this.struct_constructors.get(&def_id).cloned() {
-                                if is_expected(ctor_def) && !this.is_accessible(ctor_vis) {
-                                    err.span_label(span, format!("constructor is not visible \
-                                                                   here due to private fields"));
-                                }
+                    (Def::Struct(def_id), _) if ns == ValueNS && is_struct_like(def) => {
+                        if let Some((ctor_def, ctor_vis))
+                                = this.struct_constructors.get(&def_id).cloned() {
+                            let accessible_ctor = this.is_accessible(ctor_vis);
+                            if is_expected(ctor_def) && !accessible_ctor {
+                                err.span_label(span, format!("constructor is not visible \
+                                                              here due to private fields"));
+                            } else if accessible_ctor {
+                                let block = match ctor_def {
+                                    Def::StructCtor(_, CtorKind::Fn) |
+                                    Def::VariantCtor(_, CtorKind::Fn) => "(/* fields */)",
+                                    Def::StructCtor(_, CtorKind::Fictive) |
+                                    Def::VariantCtor(_, CtorKind::Fictive) => {
```

Comment:
Variants ctors are unreachable here and fictive struct ctor is impossible.

Reply:
Could you also add a test for this suggestion?

Reply:
So I'm is it correct for me to understand that in this case I should be checking for `ctor_def`` being `Def::Struct(..)` instead of `Def::StructCtor(_, Fictive)`?